### ISSUE TITLE
How to change the TFJob termination policy to waiting all workers to complete

### ISSUE THREAD

**Issue by 372046933** [kubeflow/kubeflow#1796]
In distributed TF training,  each worker could have a shard of the original dataset. Since the  processing speed of each worker is slightly different. When master finished, the other workers might be still reading their dataset. The default termination policy kills all other workers when master exits. When the training is resumed, all workers read the dataset from beginning again. Thus changed the distribution of input. Can I specify the termination criteria?

**Comment by jlewi** [comment]
Can you explain a little bit more about your use case? e.g. why you want to wait for all workers to exit?

We support two exit modes

1. If you have a chief; TFJob exits when the chief exits
1. If you don't have a chief we treat worker 0 as the chief and exit when worker 0 exits.

So typically you would define the appropriate termination logic in the chief.

In most cases, the chief will run for some number of global steps and exit when that number of steps reached. Or you might try to run until converged by looking at evaluation metrics.

If you are running in asynchronous training then why do you care about straggler? each worker should be processing a random shard of the data. So if some workers do more updates than other workers that shouldn't matter. You would still typically exit once the number of global steps reaches the desired values.

If you are doing synchronous training then each worker will have performed the exact same number of steps. So again you can look at global steps or number of steps performed by worker 0 or chief.

**Comment by 372046933** [comment]
I want all workers to exit normally even in asynchronous scenario. The main reason is that I am doing hyper-parameter tuning now. Every worker is reading a shard of the data-set, I want to assure that all workers have read their shard before exit. So next time, when the tuning is resumed, all workers read their shard from the beginning.  But if some worker exited halfway due to the exit of master, the tuning process will have bias against the last part of the data-set.

**Comment by 372046933** [comment]
@yuefengz If those stragglers were not checkpointed, when training resumed next time it will read the data-set from beginning. When such scenario happens again and again, the model may be over-fitted. It was common in hyper tuning process

**Comment by yuefengz** [comment]
@372046933 That sounds like a feature request to checkpoint dataset states. Maybe file an issue in tensorflow repo?

**Comment by 372046933** [comment]
Thank you very much. I can understand your point of view now. Even without kubeflow, when chief worker exited, all other workers' gradients **were not checkpointed** either.

**Comment by jlewi** [comment]
> Every worker is reading a shard of the data-set, I want to assure that all workers have read their shard before exit. So next time, when the tuning is resumed, all workers read their shard from the beginning. But if some worker exited halfway due to the exit of master, the tuning process will have bias against the last part of the data-set.

Can't you solve this by not having the chief exit until all workers completed? 

It sounds like the issue is constructing the appropriate exit-criterion for the chief. In your case, it sounds like you want all workers to do a complete pass through the data before it terminates.

There's lots of ways you could instrument this.

For example, you could have each worker write a sentinel file when its done and the chief could keep waiting until all workers write that shard.

**Comment by 372046933** [comment]
Thank you. It can be solved by manual checking.
